tests/tensor/test_nlinalg.py

```diff
@@ -214,6 +214,51 @@ def validate_shape(self, shape, compute_uv=True, full_matrices=True):
             outputs = [outputs]
         self._compile_and_check([A], outputs, [A_v], self.op_class, warn=False)
 
+    @pytest.mark.parametrize(
+        "compute_uv, full_matrices",
+        [(True, False), (False, False), (True, True)],
+        ids=[
+            "compute_uv=True, full_matrices=False",
+            "compute_uv=False, full_matrices=False",
+            "compute_uv=True, full_matrices=True",
+        ],
+    )
+    @pytest.mark.parametrize(
+        "shape", [(3, 3), (4, 3), (3, 4)], ids=["(3,3)", "(4,3)", "(3,4)"]
+    )
+    @pytest.mark.parametrize(
+        "batched", [True, False], ids=["batched=True", "batched=False"]
+    )
+    def test_grad(self, compute_uv, full_matrices, shape, batched):
+        rng = np.random.default_rng(utt.fetch_seed())
+        if batched:
+            shape = (4, *shape)
+
+        A_v = self.rng.normal(size=shape).astype(config.floatX)
+        if full_matrices:
+            with pytest.raises(
+                NotImplementedError,
+                match="Gradient of svd not implemented for full_matrices=True",
+            ):
+                U, s, V = svd(
+                    self.A, compute_uv=compute_uv, full_matrices=full_matrices
+                )
+                pytensor.grad(s.sum(), self.A)
+
+        elif compute_uv:
+            utt.verify_grad(
```

Comment:
I think verify_grad works fine for multi output expressions (you don't need to pass an Op). It would be interesting to test the following cases:

```python
def svd_uv(A):
  return svd(A, compute_uv=True)[0]
  return svd(A, compute_uv=True)[1]
  return svd(A, compute_uv=True)[2]
   
  U, s, V = svd(A, compute_uv=True)
  return U.sum() + s.sum() + V.sum()
  return U.sum() + s.sum()
  return U.sum() + V.sum()
  return s.sum() + V.sum()
```

You don't need to tweak the `verify_grad` at all if you use this approach instead of passing a single Op

Reply:
Here is an example:

```python
from functools import partial

import numpy as np
import pytensor.tensor as pt
from pytensor.gradient import verify_grad

def fn(x, case):
    max, argmax = pt.max_and_argmax(x)
    if case == 0:
        return max.sum()
    if case == 1:
        return argmax.sum()
    return max.sum() + argmax.sum()

test_x = np.random.normal(size=(5,))
verify_grad(partial(fn, case=0), pt=[test_x], rng=np.random.default_rng(0))
verify_grad(partial(fn, case=1), pt=[test_x], rng=np.random.default_rng(0))
verify_grad(partial(fn, case=2), pt=[test_x], rng=np.random.default_rng(0))
```

Reply:
Circling back to this to try your suggestion. I am running a test like this:

```
            def svd_fn(A):
                U, s, V = svd(A, compute_uv=compute_uv, full_matrices=full_matrices)
                return U.sum()

            utt.verify_grad(
                svd_fn,
                [A_v],
                rng=rng,
            )
```

But for the `input_grads` to the `L_op` I get `DisconnectedType` for any variable I'm not returning. For example, in this case I'd have `Matrix, DisconnectedType, DisconnectedType`. This raises an error because I can e.g. transpose the DisconnectedType objects. If I use all the outputs in the test, e.g. `return U.sum() + s.sum() + V.sum()`, all tests pass. Not sure how to proceed.

Reply:
Also I undid the changes to `utt.verify_grad`

Reply:
to answer your question I need to know whether the L_op should be able to handle disconnected output grads. Does it make sense to request the grad wrt to the inputs if only some of the outputs are part of the cost?

If so the failure you're seeing highlights a bug/NotImplementedError in your implementation.

If not, you should still explicitly raise a ValueError saying the grad cannot be mathematically computed when not all outputs are connected to the cost

Reply:
I guess it's case 2. If the user is only using the singular values in the cost function, `self.compute_uv` should be `False`, and we should only compute the one gradient (the first branch of the `L_op`). If he asks for all the components, he needs to use them.

Reply:
Any intuition why he needs to use them? 

My guess would be that the compute_uv case should be exactly the same whether the Op was outputting them or not, as long as the values don't change because they were outputted.

Reply:
Nope. I'm not nearly well versed enough in this stuff to have a sense of which outputs are used, when, and why. But just looking at the formulas, it seems like there are two cases: you only need the singular values (then compute_uv = False and the gradients are fine), or you need everything (then compute_uv is True and, assuming you used everything, the gradients are fine).

The other option would be to check if there are DisconnectedInputs and replace them with zero matrices of the same shape? If the matrix isn't used, the reverse mode gradients should just be all zero right?

Reply:
I'm thinking very simple here, but imagine you have an Op that returns f(x)=x, or f(x)=(x,x).

If the second output is not used the gradient with respect to x is the same in both cases, otherwise it's the sum of the two output grads in the second. 

It shouldn't matter whether there's a flag at the Op level to switch between the two versions or if the user simply ignored the second output.

I think the way I would implement this would be to check if I am getting disconnected types as output_grads and build the gradient accordingly. Setting to zeros_like(x) would be a correct solution in this case (unless both output_grads were disconnected, but I think L_op is not even called in that case)

Reply:
So long story short I don't know if setting the matrices to zeros does the right thing in your case, but I think the grad should be the same whether compute_uv is True or False and the matrices are not a part of the function cost

Reply:
MaxAndArgmax may ve useful here: https://github.com/pymc-devs/pytensor/blob/eb18f0ea4d7cc5592149f50780c4d3f6c021e90c/pytensor/tensor/math.py#L291

It explicitly handles mixed cases

Reply:
They should be variables whose type is a DisconnectedType?

Reply:
Nope, they are just `pytensor.graph.basic.Variable`

Reply:
Yeah that's what I meant, I see how just writing type didn't help much haha